Add optimized gif support for decode_gif

yongtang opened the pull request:
While revisiting the issue of #15838, I noticed that currently optimized gif is not supported. However,
optimized gif is actually possible to be processed as essentially the subsequent frame just adds the content on top of the previous frame on canvas.

This fix adds the support for optimized gif with decode_gif.

As is shown in the added test case, optimized gif (`optimized.gif`) could be handled the same way as original gif (`scan.gif`).

This fix fixes #15838.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add optimized gif support for decode_gif

While revisiting the issue of 15838, I noticed that
currently optimized gif is not supported. However,
optimized gif is actually possible to be processed
as essentially the subsequent frame just adds
the content on top of the previous frame on canvas.

This fix adds the support for optimized gif with decode_gif.

As is shown in the added test case, optimized gif (`optimized.gif`)
could be handled the same way as original gif (`scan.gif`).

This fix fixes 15838.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Format gif_io.cc with clang-format -i --style=Google

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case to cover optimized gif support for decode_gif.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

jart commented on tensorflow/core/lib/gif/gif_io.cc:
```diff
@@ -89,23 +89,52 @@ uint8* Decode(const void* srcdata, int datasize,
+      imgRight = std::min(imgRight, width);
+      imgBottom = std::min(imgBottom, height);
+
+      uint8* last_dst = dstdata + (k - 1) * width * channel * height;
```
This is an 100% optional suggestion—probably not worth doing—of which I have little understanding. But since matrix math is happening here, and "optimized" is in the PR title, just want to point out `Tensor` (based on Eigen) and better yet MKL e.g. [mkl_matmul_op.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/mkl_matmul_op.cc) (which unlike Eigen doesn't appear to be available in TF builds 100% of thte time) just in case you weren't already aware.

jart replied:
Further note: We have some code in tensorflow/contrib/tensorboard/db/ that might duplicate this code, except probably not? since I think that only applies to encode and not decode.

rmlarsen replied:
I'm not sure what TF builds @jart is referring to. I have a hard time imagining a TF build without Eigen, since many of the core classes like ```Tensor``` are thin wrappers over Eigen classes.

In any case, I don't see any matrix math with mentioning here.

yongtang committed:
Add `#include <algorithm>` to fix Windows build errors

This commit add `#include <algorithm>` to fix
build errors on Windows platform.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>
